### PR TITLE
Full repo review

### PR DESCRIPTION
**Pull request by ulrikpedersen**
Attempting to create a review of the entire repo in one go with all files added.

(probably following the same recipe as PR #6 but on that one I can only see some files added, not all)

### CODE REVIEW

File: omroneipApp/src/drvOmroneip.cpp [hunk 1]
```diff
@@ -0,0 +1,1882 @@
...
+          {
+            int bytes = 4;
+            uint8_t *rawData = (uint8_t *)malloc((size_t)(uint8_t)bytes);
+            status = plc_tag_get_raw_bytes(x.second->tagIndex, offset, rawData, bytes);
+            if (status !=0){
+              asynPrint(pasynUserSelf, ASYN_TRACE_ERROR, "%s:%s Tag index: %d Error occured in libplctag while accessing DWORD data: %s\n", driverName, functionName, x.second->tagIndex, plc_tag_decode_error(status));
+              continue;
+            }
+            epicsInt8 *pData = (epicsInt8 *)malloc(bytes * sizeof(epicsInt8));
+            /* We flip around the hex numbers to match what is done in the PLC */
```
Comment (ajgosl) [inline omroneipApp/src/drvOmroneip.cpp:1410]: In this and the following code sections, you have a lot of malloc and free calls.  It looks like the number of bytes you allocate is often fixed for non array types.  Might it be worth performing single malloc calls outside the loop and then re-using the memory wherever possible.  So for many of the cases you could probably allocate 8 bytes (or similar) and then just use however many bytess you need.
This would perhaps reduce the number of CPU cycles if you had a large database with lots of requests running at high speed.  It would also help to reduce memory fragmentation and make memory tracking easier.
For the arrays, (and this is more complicated), you could also create a storage class to hold memory of varying sizes.  Whenever you need memory ask the storage class if it has already allocated this amount and re-use.  Because I'm assuming once the records have been setup they will always ask for the same qty of items?  This would then definitely reduce memory malloc/frees, in fact eliminate them altogether.  You would simply free everything at the end when you break out of the loop.

File: omroneipApp/src/drvOmroneip.cpp [hunk 2]
```diff
@@ -0,0 +1,1882 @@
...
+asynStatus drvOmronEIP::readInt8Array(asynUser *pasynUser, epicsInt8 *value, size_t nElements, size_t *nIn)
+{
+  const char * functionName = "readInt8Array";
+  int status;
+  for (auto x : tagMap_)
+  {
+    int offset = x.second->tagOffset;
+    if (x.second->dataType == "UDT")
+    {
+      size_t bytes = plc_tag_get_size(x.second->tagIndex);
```
Comment (ajgosl) [inline omroneipApp/src/drvOmroneip.cpp:1509]: Same memory storage class idea to eliminate constant malloc/free

**Review by ajgosl — CHANGES_REQUESTED** [review]
- Builds fine.  I'm guessing you haven't started on the docs yet so a little bit on building libplctag etc when you do get around to them would be good.  But it wasn't too hard to get everything built and running.
- Could run up the individual test IOCs and then interact with the records.  They worked consistently, although I couldn't verify the value in the simulator had changed (this is an issue with the libplctag simulator, not this module).

File: iocBoot/iocTest/cdCommands
Comment (ulrikpedersen) [inline iocBoot/iocTest/cdCommands:1]: Delete this file from the repo. It is a generated product. Also add to `.gitignore`.

File: iocBoot/iocTest/asynTrace.out
Comment (ulrikpedersen) [inline iocBoot/iocTest/asynTrace.out:1]: Delete this file from the repo. It is a generated product. Also add to `.gitignore`.

File: iocBoot/iocTest/envPaths
Comment (ulrikpedersen) [inline iocBoot/iocTest/envPaths:1]: Delete this file from the repo. It is a generated product. Also add to `.gitignore`.

**Review by ulrikpedersen — COMMENTED** [review]
OK, so this was just a first-round of review with the lowest-hanging fruit 🍎  - the structure and content of the support module. 

There is a bunch of stuff that has snuck in which don't really belong in the repo. We need to remove these files/dirs and improve the `.gitignore` to catch these and ignore them in future.

I'm going to look at the C++ code now. That might take a while as I'm rusty...

**Comment by ulrikpedersen** [comment]
I don't see a `LICENSE` file on this repo, but it is Public so right now anyone can grab it and claim it as their own... or something... 

Do we know what license this should eventually be? If not, then perhaps we should slap something on it until we can hand it over - since we own the copyright entirely (only OSL contributions) we can always change it later to whatever the client want.

**Review by ulrikpedersen — COMMENTED** [review]
> There are printf statements, std::cout and asynPrint.  Can these be made consistent.  Ideally, no printf or cout unless perhaps right at the beginning if the driver fails. 

Definitely go with `asynPrint` as soon as the asyn port is created :+1:

File: omroneipApp/src/drvOmroneip.cpp [hunk 3]
```diff
@@ -0,0 +1,1882 @@
...
+{
+  drvOmronEIP *pPvt = (drvOmronEIP *)drvPvt;
+  pPvt->readPoller();
+}
+
+// this thread runs once after iocInit to optimise the tag map before the thread poller begins
+static int optimiseTagsC(void *drvPvt)
+{
+  drvOmronEIP *pPvt = (drvOmronEIP *)drvPvt;
+  while (!iocStarted)
```
Comment (ulrikpedersen) [inline omroneipApp/src/drvOmroneip.cpp:63]: At a glance and judging by variable name `iocStarted` I suspect you're waiting for `iocInit()` to complete. However, there are hooks/events for the various stages of IOC initialisation which I would recommend instead of a polling loop. See https://docs.epics-controls.org/en/latest/appdevguide/IOCInit.html#inithooks The hook you're after is probably `initHookAfterIocRunning`

File: omroneipApp/src/drvOmroneip.h [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+  drvOmronEIP(const char *portName,
+              char *gateway,
+              char *path,
+              char *plcType,
+              int debugLevel);
+  ~drvOmronEIP();
+  bool omronExiting_;
+
+  void readPoller();
+  /* Takes a csv style file, where each line contains a structure name followed by a list of datatypes within the struct
```
Comment (ulrikpedersen) [inline omroneipApp/src/drvOmroneip.h:82]: I would suggest moving these comments to the function in the .cpp file and format them as doxygen. That seems to be most common place to put this although Doxygen *might* support docs-comments in both header and source...

File: omroneipApp/src/drvOmroneip.cpp (hunk 3 above)
Comment (ulrikpedersen) [reply omroneipApp/src/drvOmroneip.cpp:63]: Ah, I see now that you *are* using the initHook :+1:

Then what I would suggest is to use an epicsEvent to signal that you're started/initialized instead of going around a while loop. The epicsEvent.h header has this API and the old Application Developers Guile describe it on p. 289 (I couldn't easily find it in the modern EPICS docs) 

https://epics.anl.gov/base/R3-14/12-docs/AppDevGuide.pdf

File: omroneipApp/src/drvOmroneip.cpp [hunk 5]
```diff
@@ -0,0 +1,1882 @@
...
+  initHookRegister(myInitHookFunction);
+  asynStatus status = (asynStatus)(epicsThreadCreate("optimiseTags",
+                          epicsThreadPriorityMedium,
+                          epicsThreadGetStackSize(epicsThreadStackMedium),
+                          (EPICSTHREADFUNC)optimiseTagsC,
+                          this) == NULL);
+  if (status!=0){
+    std::cout<< "Error, Driver initiation returned asyn status:" << status << std::endl;
+  }
+  initialized_ = true;
```
Comment (ulrikpedersen) [inline omroneipApp/src/drvOmroneip.cpp:129]: Send an epicsEvent signal here instead

File: omroneipApp/src/drvOmroneip.cpp [hunk 6]
```diff
@@ -0,0 +1,1882 @@
...
+}
+
+asynStatus drvOmronEIP::createPoller(const char *portName, const char *pollerName, double updateRate)
+{
+  int status;
+  omronEIPPoller* pPoller = new omronEIPPoller(portName, pollerName, updateRate);
+  pollerList_[pPoller->pollerName_] = pPoller;
+  status = (epicsThreadCreate(pPoller->pollerName_,
```
Comment (ulrikpedersen) [inline omroneipApp/src/drvOmroneip.cpp:137]: There is also an epicsThread C++ interface which might make this code look a little bit more like "proper" C++ ;-) 

It just wraps the C interface and so there is no real technical reason to change. Just aestethics if you're into that...

File: omroneipApp/src/drvOmroneip.cpp [hunk 7]
```diff
@@ -0,0 +1,1882 @@
...
+  }
+
+  // Get the required data from the drvInfo string
+  std::unordered_map<std::string, std::string> keyWords = drvInfoParser(drvInfo);
+  if (keyWords.at("stringValid") != "true")
+  {
+    readFlag = false;
+    printf("Error! drvInfo string is invalid, record: %s was not created\n", drvInfo);
+    asynPrint(pasynUserSelf, ASYN_TRACE_ERROR, "%s:%s drvInfo string is invalid, record: %s was not created\n", driverName, functionName, drvInfo);
+    tag = "Invalid tag!";
```
[the remark this answers is not in the document]
Comment (ulrikpedersen) [reply omroneipApp/src/drvOmroneip.cpp:183]: and a `cout` 10 lines later! Now we just need an `epicsPrintf` and `errLogPrintf` to achieve a Royal Flush of logging :-D

[the remark this answers is not in the document]
Comment (ulrikpedersen) [reply omroneipApp/src/drvOmroneip.cpp:183]: Sorry, its Friday and I'm getting silly from reviewing...

File: omroneipApp/src/drvOmroneip.cpp [hunk 8]
```diff
@@ -0,0 +1,1882 @@
...
+  {
+    asynPrint(pasynUserSelf, ASYN_TRACE_FLOW, "%s:%s Asyn param: %d optimisationFlag: '%s' readFlag: %s\n", driverName, functionName, tag.first, tag.second->optimisationFlag.c_str(), tag.second->readFlag ? "true" : "false");
+  }
+
+  this->unlock();
+  this->startPollers_ = true;
+  return asynSuccess; // always return success as any failures to optimise should leave the driver in its pre omtimised state which should have no erros
+}
+
+std::unordered_map<std::string, std::string> drvOmronEIP::drvInfoParser(const char *drvInfo)
```
Comment (ulrikpedersen) [inline omroneipApp/src/drvOmroneip.cpp:528]: This is a *really* long function and kind of hard to follow from end-to-end. However, it does seem to have different sections. Consider breaking these sections down into smaller (private) functions?

Comment (ulrikpedersen) [reply omroneipApp/src/drvOmroneip.cpp:528]: Maybe even create a smaller, stand-alone (unit-testable) class that just has the job to parse drvInfo strings...

File: omroneipApp/src/drvOmroneip.cpp [hunk 9]
```diff
@@ -0,0 +1,1882 @@
...
+        keyWords.at("offset") = std::to_string(offset);
+      }
+      words.pop_front();
+    }
+    else if (i == 4)
+    {
+      // Check for valid extra attributes
+      /* These attributes overwrite libplctag attributes, other attributes which arent overwritten are not mentioned here
+        Users can overwrite these defaults and other libplctag defaults from their records */
+      std::unordered_map<std::string, std::string> defaultTagAttribs = {
```
Comment (ulrikpedersen) [inline omroneipApp/src/drvOmroneip.cpp:799]: This `unordered_map<std::string, std::string>` repeats in a lot of places. Could we either typedef it or perhaps wrap it in a class/object?

File: omroneipApp/src/drvOmroneip.cpp [hunk 10]
```diff
@@ -0,0 +1,1882 @@
...
+  {
+    asynPrint(pasynUserSelf, ASYN_TRACE_FLOW, "%s \t\t\t %s\n", i->first.c_str(), i->second.c_str());
+  }
+  return keyWords;
+}
+
+asynStatus drvOmronEIP::loadStructFile(const char * portName, const char * filePath)
+{
+  const char * functionName = "loadStructFile";
+  this->lock();
```
Comment (ulrikpedersen) [inline omroneipApp/src/drvOmroneip.cpp:887]: Do we need to lock the asynport in this function?

File: omroneipApp/src/drvOmroneip.cpp [hunk 11]
```diff
@@ -0,0 +1,1882 @@
...
+      }
+      else {
+        // Padding may be required at the end of the structure
+        nextDtype=parentNextDtype;
+      }
+    }
+
+    if (dtype =="UINT" || dtype =="INT" || dtype =="WORD" || dtype =="BOOL")
+      dtypeSize=2;
+    else if (dtype =="UDINT" || dtype =="DINT" || dtype =="REAL" || dtype =="DWORD")
```
Comment (ulrikpedersen) [inline omroneipApp/src/drvOmroneip.cpp:1025]: are `UINT` and `INT` really only 16bits? Presumably that's the PLCs datatypes...?

File: omroneipApp/src/drvOmroneip.cpp [hunk 12]
```diff
@@ -0,0 +1,1882 @@
...
+              for (size_t m = 0; m<offsetSubstring.size(); m++)
+              {
+                if (offsetSubstring.c_str()[m] == ']')
+                {
+                  try
+                  {
+                    //struct integer found
+                    structIndex = std::stoi(offsetSubstring.substr(0,m));
+                    keyWords.at("optimisationFlag") = "attempt optimisation";
+                    goto findOffsetFromStruct;
```
Comment (ulrikpedersen) [inline omroneipApp/src/drvOmroneip.cpp:747]: oh a `goto` - that's a blast from the past. I think these should *not* be used as part of the expected code path - the risk of subtle bugs creeping into future modifications is too great. Please could you think of refactoring this snippet to avoid the goto?

Within the `for` loop you could for example use `break` instead.

File: omroneipApp/src/drvOmroneip.cpp [hunk 13]
```diff
@@ -0,0 +1,1882 @@
...
+                    std::cout << "Error, could not find a valid index for the requested structure: " << words.front() << std::endl;
+                    keyWords.at("stringValid") = "false";
+                  }
+                }
+              }
+            }
+          }
+          std::cout<<"Invalid index for requested structure: " << words.front() << std::endl;
+          keyWords.at("stringValid") = "false";
+
```
Comment (ulrikpedersen) [inline omroneipApp/src/drvOmroneip.cpp:760]: These lines really appear to be an error handler - but without being explicit about it because the `goto` above just normally skips over this bit.
It would make for more readable code to have this explicitly branched as an error handler, for example in an `if` statement.

Comment (ulrikpedersen) [reply omroneipApp/src/drvOmroneip.cpp:760]: In some 25ish lines below there is an example of the sort of clearly defined error handler I'm thinking of. `if(!structFound)`... :+1:

File: omroneipApp/src/drvOmroneip.cpp [hunk 14]
```diff
@@ -0,0 +1,1882 @@
...
+        bool intFound = false;
+        for (size_t i = 0; i<ss.size(); i++)
+        {
+          if (ss.substr(i,1)== "]")
+          {
+            intFound = true;
+            try
+            {
+              strLength = std::stoi(ss.substr(0,i+1));
+              if (strLength < 0) throw 1;
```
Comment (ulrikpedersen) [inline omroneipApp/src/drvOmroneip.cpp:1096]: I think this is the only place where we throw and exception (and the exception is just 1???)
For the rest of this function, errors are indicated by returning -1. Its best to be consistent in the error-handling - definitely within the same function, within the class and ideally within the module, although there may be ... exceptions (pun fully intended!) to this rule at a module level.

File: omroneipApp/src/drvOmroneip.cpp [hunk 15]
```diff
@@ -0,0 +1,1882 @@
...
+{
+  static const char *functionName = "readPoller";
+  std::string threadName = epicsThreadGetNameSelf();
+  std::string tag;
+  int offset;
+  int status;
+  int still_pending = 1;
+  int timeTaken = 0;
+  //The poller may not be fully initialised until the startPollers_ flag is set to 1, do not attempt to get the pPoller yet
+  while (!this->startPollers_ && !omronExiting_) {epicsThreadSleep(0.1);}
```
[the remark this answers is not in the document]
Comment (ulrikpedersen) [reply omroneipApp/src/drvOmroneip.cpp:1214]: I eventually caught up with Alan here... I agree with this one. I would definitely use epicsEvents over while(sleep) loops.

File: omroneipApp/src/drvOmroneip.cpp [hunk 16]
```diff
@@ -0,0 +1,1882 @@
...
+  omronEIPPoller* pPoller = pollerList_.at(threadName);
+  double interval = pPoller->updateRate_;
+  asynPrint(pasynUserSelf, ASYN_TRACE_FLOW, "%s:%s Starting poller: %s with interval: %f\n", driverName, functionName, threadName.c_str(), interval);
+  while (!omronExiting_)
+  {
+    double waitTime = interval - ((double)timeTaken / 1000);
+    if (waitTime >=0) {
+      epicsThreadSleep(waitTime);
+    }
+    else {
```
Comment (ulrikpedersen) [inline omroneipApp/src/drvOmroneip.cpp:1224]: This bit here is another candidate for using epicsEvent. I would create an epicsEvent called `omronExiting_` and change the while loop to `while(true)`. Then instead of the `epicsThreadSleep` I would use a the `omronExiting_.wait(waitTime)` (probably works also with waitTime = 0) and IF the event has been signalled, just `break` out of the while loop.

Comment (ulrikpedersen) [reply omroneipApp/src/drvOmroneip.cpp:1224]: https://epics.anl.gov/base/R3-14/12-docs/AppDevGuide.pdf p.289

File: omroneipApp/src/drvOmroneip.cpp [hunk 17]
```diff
@@ -0,0 +1,1882 @@
...
+              setParamStatus(x.first, asynTimeout);
+              readFailed = true;
+              asynPrint(pasynUserSelf, ASYN_TRACE_FLOW, "%s:%s Timeout finishing read tag %d: %s. Try decreasing the polling rate\n", driverName, functionName, x.second->tagIndex, plc_tag_decode_error(status));
+              break;
+            }
+            else
+            {
+              still_pending = 1;
+              waits++;
+              epicsThreadSleep(0.01);
```
Comment (ulrikpedersen) [inline omroneipApp/src/drvOmroneip.cpp:1269]: Why is this sleep 0.01 needed here?

File: omroneipApp/src/drvOmroneip.cpp (hunk 1 above)
Comment (ulrikpedersen) [reply omroneipApp/src/drvOmroneip.cpp:1410]: I like the idea of a wrapper class with RAII - free all the memory when going out of scope.

File: omroneipApp/src/drvOmroneip.cpp (hunk 12 above)
Comment (ptsOSL) [reply omroneipApp/src/drvOmroneip.cpp:747]: I distinctly remember feeling very smart when I used that goto and that there was some reason that I couldnt use a break. I think i probably just didn't want to create another flag like indexFound to put around the debugging output and thought this was a neater, although harder to follow, approach.

File: omroneipApp/src/drvOmroneip.cpp (hunk 11 above)
Comment (ptsOSL) [reply omroneipApp/src/drvOmroneip.cpp:1025]: Yes they are

**Comment by ulrikpedersen** [comment]
OK, I think that's me done code-reviewing 😅 

In summary my above comments can probably be distilled down to a few points:
* use asynPrint in the asyn driver and avoid printf, cout, etc. unless you have no other choice (i.e. before asyn port is created)
* Use `epicsEvent` to signal/wait for synchronisation between threads instead of `while(boolean)`
* Consistent error handling: either return error codes OR throw exception, not both. Bonus point for error codes that have meaning...


Finally, I think a refactor is in order: right now everything is the asynPortDriver class. However, many functions really only deal with handling, parsing, or in some form working with the PLC tags and have no EPICS or asyn in them. These would benefit from being pulled out into a separate class `PlcTagParsingAndManipulationOrSomething` - which could then benefit from going through unittesting since these functions are nontrivial.

I'm thinking these functions might all not need to be in the asynPortDriver class:
```
  asynStatus loadStructFile(const char * portName, const char * filePath);
  asynStatus createStructMap(std::unordered_map<std::string, std::vector<std::string>> rawMap);
  std::vector<std::string> expandStructs(std::unordered_map<std::string, std::vector<std::string>> const& rawMap, std::string structName);
  size_t findOffsetsRecursive(std::unordered_map<std::string, std::vector<std::string>> const& rawMap, std::unordered_map<std::string, std::vector<std::string>> const& expandedMap, std::string structName, std::unordered_map<std::string, std::vector<int>>& structMap, std::string parentNextDtype);
  std::string findNextDtype(std::unordered_map<std::string, std::vector<std::string>> const& rawMap, std::string structName);
  std::unordered_map<std::string, std::string> drvInfoParser(const char *drvInfo);
```

File: omroneipApp/src/drvOmroneip.h
[the remark this answers is not in the document]
Comment (ptsOSL) [reply omroneipApp/src/drvOmroneip.h:1]: I have made a ticket for this, I will do if I get time

File: configure/RELEASE [hunk 18]
```diff
@@ -0,0 +1,19 @@
...
+# Run "gnumake clean uninstall install" in the application top
+# directory whenever this file is changed or any external module
+# pointed to by this RELEASE file is rebuilt.
+
+# If you don't want to install files into $(TOP) then define
+# INSTALL_LOCATION_APP here:
+#INSTALL_LOCATION_APP=<fullpathname>
+
+# Asyn
+ASYN = /home/epics-dev/work/epics/modules/asyn/asyn-R4-44-2
```
[the remark this answers is not in the document]
Comment (ptsOSL) [reply configure/RELEASE:12]: That is my understanding, ill leave it as is

File: omroneipApp/src/drvOmroneip.cpp (hunk 13 above)
Comment (ptsOSL) [reply omroneipApp/src/drvOmroneip.cpp]: Reformatted

File: omroneipApp/src/drvOmroneip.cpp (hunk 14 above)
Comment (ptsOSL) [reply omroneipApp/src/drvOmroneip.cpp]: I dont think it really matters here what I throw, we don't print the thrown exception, rather printing the same sentence for any of the three possible causes. The main purpose of this try catch is to catch exceptions from stoi, by throwing on the if (arrayStart ... line, I save some code rather than having another flag to check. I could expand the logging by checking the exception and acting upon it, but i dont think it is necessary here.

 I can throw -1 instead though I suppose

File: omroneipApp/src/drvOmroneip.cpp (hunk 17 above)
Comment (ptsOSL) [reply omroneipApp/src/drvOmroneip.cpp:1509]: I have essentially implemented by own timer here. I dont want to progress until a timeout has expired and am using a combination of counting up waits and epicsThreadSleep to count up to the timeout. I think I did it this way rather than allowing the thread to constantly loop and checking the actual time, as I thought it was useful to sleep the thread. This was maybe back when the thread was locking? Anyway, I will switch it to look at a clock which is more intuitive

Comment (ptsOSL) [reply omroneipApp/src/drvOmroneip.cpp]: I have changed it to use a std::chrono timer and I think it should be fine. It is a bit complex as there is both the timing interval for the poller and the timeout for the record. We send all the read requests at the start. This loop we are looking at is waiting for timeouts if the status we recieve from libplctag is PLCTAG_STATUS_PENDING. It will wait until this records timeout is expired and then skip to the next item in the poller. Hopefully this item will now have been fetched, if not then if its timeout was the same as the previous record, it will now have expired etc. If the user sets timeouts so that they extend beyond the read interval requested, then this is reported, however we still wait for the timeouts and do not abort this poll.

Comment (ptsOSL) [reply omroneipApp/src/drvOmroneip.cpp]: I have just pushed the changes now

File: omroneipApp/src/drvOmroneip.cpp (hunk 9 above)
Comment (ptsOSL) [reply omroneipApp/src/drvOmroneip.cpp:793]: I have implemented:
typedef std::unordered_map<std::string, std::vector<std::string>> structDtypeMap;
typedef std::unordered_map<std::string, std::string> drvInfoMap;

To tidy up the two worst offenders.

File: omroneipApp/src/drvOmroneip.cpp (hunk 10 above)
Comment (ptsOSL) [reply omroneipApp/src/drvOmroneip.cpp]: Im pretty sure this is no longer needed, so have removed it. I needed to ensure that this was run before the database files are loaded and not at the same time. If they do run at the same time then there can be a seg fault, but I dont think the lock is required for this. This should happen naturally. I have also added a check to make sure that this function can only be run before the ioc has started running.

File: omroneipApp/src/drvOmroneip.cpp [hunk 19]
```diff
@@ -0,0 +1,1882 @@
...
+}
+
+void drvOmronEIP::readPoller()
+{
+  static const char *functionName = "readPoller";
+  std::string threadName = epicsThreadGetNameSelf();
+  std::string tag;
+  int offset;
+  int status;
+  int still_pending = 1;
```
[the remark this answers is not in the document]
Comment (ptsOSL) [reply omroneipApp/src/drvOmroneip.cpp:1451]: What do you mean by magic numbers? I have changed it though, i have no idea why it was an int, presumably it could have different values at some point in the past.

File: omroneipApp/src/drvOmroneip.cpp (hunk 3 above)
Comment (ptsOSL) [reply omroneipApp/src/drvOmroneip.cpp:59]: The reason iocStarted was not an epicsEvent, is because either the iocStarted flag or the iocStartedEvent would have to be global due to the required format of the hook function. (static void myInitHookFunction(initHookState state))

Looking at it now however, I dont think there is really any difference between having a global bool and a global epicsEvent, so I will change this.

Comment (ptsOSL) [reply omroneipApp/src/drvOmroneip.cpp:59]: Is there any way of finding out if an event has fired in the past? Elsewhere in the code I need to check if the event has fired in the past

Comment (ptsOSL) [reply omroneipApp/src/drvOmroneip.cpp:59]: After I have triggered and then waited for the signal elsewhere, I could potentially rearm the event and then check that the event status is full? This would mean that the signal had either been fired and received once or just fired. It doesnt feel great though.

File: omroneipApp/src/drvOmroneip.cpp (hunk 5 above)
Comment (ptsOSL) [reply omroneipApp/src/drvOmroneip.cpp:126]: How exactly would I do that here? I could create an event in the drivers constructor which fired an initializationEvent, then each time drvUserCreate is called I would wait for this event with a timeout? I suppose I would then have to reset the event each time it is checked for in drvUserCreate? Is there any way to leave the event in a triggered state rather rather than resetting it each time? Because refiring it for the hundreds of records I have to create wouldn't make much sense.

File: omroneipApp/src/drvOmroneip.cpp (hunk 15 above)
[the remark this answers is not in the document]
Comment (ptsOSL) [reply omroneipApp/src/drvOmroneip.cpp:1456]: Each configured poller runs in its own thread which runs the readPoller() function. So if I changed this to an event and triggered it when the driver is ready to start the pollers, only the first polling thread which checks for the event will break out of the loop. I could then use this poller to refire the event and so on until all pollers have seen it, but this would be a bit messy. Hence why I have used this flag instead which is set to true in one place only, and is checked by all of the different polling threads.

File: omroneipApp/src/drvOmroneip.cpp (hunk 1 above)
Comment (ptsOSL) [reply omroneipApp/src/drvOmroneip.cpp]: This might be quite involved, so I will create a separate branch and merge request for this later on.

File: omroneipApp/src/drvOmroneip.h (hunk 4 above)
Comment (ptsOSL) [reply omroneipApp/src/drvOmroneip.h]: I will do something like Doxygen at some point, but will create a seperate branch for it

File: omroneipApp/src/drvOmroneip.cpp (hunk 8 above)
Comment (ptsOSL) [reply omroneipApp/src/drvOmroneip.cpp]: I will refactor this and some other "non driver" related functions in a seperate branch and I do intend to have unit tests for them.

File: omroneipApp/src/drvOmroneip.cpp (hunk 16 above)
Comment (ptsOSL) [reply omroneipApp/src/drvOmroneip.cpp:1466]: This has the same problem as the startPollers_ event, each poller needs to see this event so that it can shut down.

File: omroneipApp/src/drvOmroneip.cpp [hunk 20]
```diff
@@ -0,0 +1,1882 @@
...
+        // attempt to set offset to integer, if not possible then assume it is a structname
+        try
+        {
+          offset = std::stoi(words.front());
+          keyWords.at("optimisationFlag") = "attempt optimisation";
+        }
+        catch(...)
+        {
+          // attempt to split name and integer
+          for (size_t n = 0; n<words.front().size(); n++)
```
[the remark this answers is not in the document]
Comment (ptsOSL) [reply omroneipApp/src/drvOmroneip.cpp]: At this point in the code we are looking through the individual characters of one of the words which should look something like: struct[1][4][2][42]. I need to look for each set of square brackets and get the indices within. I couldnt think of a way of doing this with iterators, I could use string:find, but would still need to keep track of indices and create substrings to find the second, third etc instance of brackets. I have improved the code comments, but dont think it is worth rewriting the code to use string:iterators exclusively.

File: omroneipApp/src/drvOmroneip.cpp (hunk 6 above)
Comment (ptsOSL) [reply omroneipApp/src/drvOmroneip.cpp:134]: I did not know about this, I will try and remember for the next time I need to use epics threading.
